File: python/medium/power_set.py
from typing import List, Set, FrozenSet

flowers = ['', 'a', 'b'];

def power_list(data: List[str]) -> Set[Set[str]]:
    fragrances = [[]]

    for flower in data:
        new_fragrances = []
        
        for fragrance in fragrances:
            new_fragrances.append(fragrance + [flower])
        fragrances = fragrances + new_fragrances
    return fragrances
print(power_list(flowers))

def power_set(data: Set[str]) -> Set[FrozenSet[str]]:
    fragrances = {frozenset()}

    for flower in data:
        new_fragrances = set()

        for fragrance in fragrances:
            new_fragrances.add(fragrance.union({flower}))
        fragrances.update(new_fragrances)
    return fragrances
print(power_set({'', 'a', 'b'}))

File: ts/medium/powerSet.ts
const flowers = ['', 'a', 'b'];

function powerArr(data: string[]): string[][] {
  let fragrances: string[][]  = [[]];


  for (let i = 0; i < data.length; i++) {
    const flower = data[i];
    const newFragrances: string[][]  = [];

    for (const fragrance of fragrances) {
      newFragrances.push([...fragrance, flower]);
    }
    fragrances = [...fragrances, ...newFragrances];
  }
  return fragrances;
}
console.log(powerArr(flowers));

function powerSet(data: string[]): Set<Set<string>> {
  let fragrances: Set<Set<string>> = new Set();
  fragrances.add(new Set());

  for (const flower of data) {
    const newFragrances: Set<Set<string>> = new Set();

    for (const fragrance of fragrances) {
      const newFragrance = new Set(fragrance);
      newFragrance.add(flower);
      newFragrances.add(newFragrance);
    }

    for (const item of newFragrances) {
      fragrances.add(item);
    }
  }

  return fragrances;
}
console.log(powerSet(flowers));